### PR TITLE
Akopyan Zal. Lab 1. Var 2.

### PR DESCRIPTION
**Pull request by akopyanzal**
I made plugin that adds the inline attribute

### CODE REVIEW

File: clang/test/lab1/akopyan_zal/Test.cpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/AlwaysInlinePluginAkopyan%pluginext -plugin always-inline %s 2>&1 | FileCheck %s
+
+// CHECK: __attribute__((always_inline)) FuncInline
```
Comment (m-ly4) [inline clang/test/lab1/akopyan_zal/Test.cpp]: Please check attribute presence in AST tree.

File: clang/test/lab1/akopyan_zal/Test.cpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/AlwaysInlinePluginAkopyan%pluginext -plugin always-inline %s 2>&1 | FileCheck %s
+
+// CHECK: __attribute__((always_inline)) FuncInline
+int FuncInline(){ return 10; }
+
```
Comment (m-ly4) [inline clang/test/lab1/akopyan_zal/Test.cpp]: Please check that attribute is absent in the following example
```suggestion
void FuncNotInline() {
  {
  if (true) {
    return;
  }
}
  return;
}
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by akopyanzal** [comment]
Why my tests are not running?

File: clang/labs/lab1/akopyan_zal/AlwaysInline.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    return std::make_unique<AlwaysInlineConsumer>(&Compiler.getASTContext());
+  }
+  bool ParseArgs(const CompilerInstance &Compiler,
+                 const std::vector<std::string> &args) override {
+    return true;
+  }
+};
+
+static FrontendPluginRegistry::Add<AlwaysInlinePlugin>
+    X("always-inline", "Plugin that adds the inline attribute");
```
Comment (m-ly4) [inline clang/labs/lab1/akopyan_zal/AlwaysInline.cpp]: Try to change dash in plugin name on underscore:
```suggestion
    X("always_inline", "Plugin that adds the inline attribute");
```

File: clang/test/lab1/akopyan_zal/Test.cpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// RUN: %clang_cc1 -ast-dump -ast-dump-filter -load %llvmshlibdir/AlwaysInlinePluginAkopyan%pluginext -plugin always-inline %s 2>&1 | FileCheck %s
```
Comment (m-ly4) [inline clang/test/lab1/akopyan_zal/Test.cpp]: Here name should be changed too:
```suggestion
// RUN: %clang_cc1 -ast-dump -ast-dump-filter -load %llvmshlibdir/AlwaysInlinePluginAkopyan%pluginext -plugin always_inline %s 2>&1 | FileCheck %s
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/test/lab1/akopyan_zal/Test.cpp (hunk 4 above)
Comment (akopyanzal) [reply clang/test/lab1/akopyan_zal/Test.cpp]: done

**Comment by m-ly4** [comment]
If you have any questions regarding this lab you may come to practice and ask them.

File: clang/labs/lab1/akopyan_zal/AlwaysInline.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  bool VisitFunctionDecl(FunctionDecl *func) {
+    Stmt *Body = func->getBody();
+    if (Body && isa<CompoundStmt>(Body)) {
+      CompoundStmt *BodyCompound = cast<CompoundStmt>(Body);
+      for (Stmt *S : BodyCompound->body()) {
+        if (!isa<IfStmt>(S)) {
+          func->addAttr(AlwaysInlineAttr::CreateImplicit(*Context));
+          llvm::outs() << "__attribute__((always_inline)) "
+                       << func->getNameAsString() << "\n";
```
Comment (m-ly4) [inline clang/labs/lab1/akopyan_zal/AlwaysInline.cpp]:
```suggestion
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/test/lab1/akopyan_zal/Test.cpp [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+// CHECK: __attribute__((always_inline)) FuncInline
+int FuncInline(){ return 10; }
+
+void FuncNotInline() {
+  if (true) {
+    return;
+  }
+  return;
+}
```
Comment (m-ly4) [inline clang/test/lab1/akopyan_zal/Test.cpp]: What about other statements (loops, switch)?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**